Fix GIL protection with CVirtualHelper

kxrob opened the pull request:
Fixes GIL issues (after 3c9271d + 27a1ae8 and pre-existing) noticed during debugging of win32ui GUI code with threads: (indirect) Python API calls outside of GIL  (E.g.  win32template.cpp / CPythonDocTemplate::InitialUpdateFrame / GetGoodRet() / DODECREF ).

Now holds GIL during CVirtualHelper instance life by default.

Also reduces GIL enter/leave overhead: Mostly only 1x
is necessary instead of 3x or more.

mhammond commented:
Sorry that this slipped through the cracks. Can you explain in more detail what problem this fixes? Given this touches so much code I'm surprised the flaws in the existing implementation aren't immediately obvious.

kxrob commented:
> Can you explain in more detail what problem this fixes? Given this touches so much code I'm surprised the flaws in the existing implementation aren't immediately obvious.

Rebased and updated descriptions.
Those recent changes  3c9271d + 27a1ae8  introduced (more) Python API calls being possible outside GIL. Corruptions happen rarely with threads + win32ui only.
The 1st main commit is small, and removes more explicit CEnterLeavePython's than adding code.
The 2nd is only for further (optional) optimizations to reduce GIL time a little more - via some early routine helper.release_full() 's.

mhammond commented on Pythonwin/ddeserver.cpp:
```diff
@@ -27,7 +27,7 @@ CDDEServerSystemTopic *PythonDDEServer::CreateSystemTopic()
     PyObject *ob;
     if (helper.call() && helper.retval(ob)) {
-        CEnterLeavePython _celp;
+        //CEnterLeavePython _celp;
```
Please delete lines rather than commenting them out (also below, but I haven't commented against them all)

kxrob replied:
Done

mhammond reviewed (changes requested):
This looks fine to me, thanks - please just remove some of the commented lines and I'll commit it.
